### PR TITLE
Run Cypress integration tests against the remote dev environment

### LINKED ISSUE #1770: Run e2e tests against deployed environments
### Description
Run e2e tests against deployed environments. This should run against an environment that's deployed against the regular flow, and uses Okta Preview and the Experian demo account. We'll need to add `dev` back to the deployment pipeline for this to work.

### Proposed solution 
After deploying to training or demo, add a step to run the e2e tests, configured to point at the dev URL

### Additional context
This ticket was created in response to [this postmortem](https://github.com/usds/prime-simplereport-docs/pull/15)

We may need to add different profiles for running the e2e tests locally vs in deployed environments. Locally we'll want to use Wiremock instead of Okta preview, for instance.

This should be run before deploying a PR to production, rather than just doing it daily. We may need to add a flag enabling users to bypass e2e tests to respond to production incidents.

### PR DESCRIPTION
**Pull request by alismx**
## Related Issue

- resolves #1770 

### CI

- Deploy main changes to stg
- Run Cypress against stg
- Block production deploys if cypress fails
- Include a way to skip running cypress against dev (include `skip e2e stg` in the latest commit message)

### Cypress

Disable video recording by default (this speeds cypress up)
Generate unique names for organizations and facilities (faker) 
Generate unique emails for users (faker)
Update tests to handle multiple facilities if needed
Update tests to navigate to pages and links for better coverage
Update Wiremock to allow for facilities with different names

### Frontend

Set an `id` on the settings link for more accessible cypress selection and navigation.
 
## Testing

- Run a deploy to dev on this branch. 
 
## Checklist for Primary Reviewer

- [ ] Does it seem like the gate for production deploys will work as intended?

### Infrastructure
- [ ] Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed
- [ ] Pipeline is secure and obfuscates any sensitive data

## Cloud
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

**Comment by alismx** [comment]
I'll be addressing the failed e2e tests that are run locally on the github runner.

**Comment by alismx** [comment]
This should be reviewable. Another PR for converting these jobs to docker is a WIP.

File: frontend/cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json
Comment (nathancrtr) [inline frontend/cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: Consider checking for email-like strings here (contain `@` and `.`)

**Review by emmastephenson — COMMENTED** [review]
I'm sorry to ask this so late in the game, but what's the rationale for using `dev` to run these tests instead of one of the environments that's already included in our prod deploys? I worry about a situation where someone is using `dev` to test out their changes and the changes are suddenly overwritten when someone else deploys to prod.

**Comment by jdorothy** [comment]
> > I'm sorry to ask this so late in the game, but what's the rationale for using `dev` to run these tests instead of one of the environments that's already included in our prod deploys? I worry about a situation where someone is using `dev` to test out their changes and the changes are suddenly overwritten when someone else deploys to prod.
> 
> Fair question! I'm open to having this run in a different environment. That shouldn't be a complex lift for this PR, mostly just making sure our cypress user has access and updating the cypress base url.
> 
> As far as `dev` usage goes, my understanding is that `dev` would no longer be available for that type of testing or use. While that isn't entirely ideal in the short term, we have a ticket to enable the creation of ephemeral environments per PR to solve potential testing capacity issues.
> 
> Which environment seems like a better candidate?

I'd suggest `stg`, since it's our prod-like environment.

File: .github/workflows/deployStg.yml
Comment (jdorothy) [inline .github/workflows/deployStg.yml:120]: Why run e2e tests twice as part of the same merge? If I'm reading this right, the tests will run in `dev` and then again in `stg` (but still against `dev`)?

Comment (alismx) [reply .github/workflows/deployStg.yml:120]: This is only running a health check and not the full suite. Per the other comments, this will just be changed to the full run.

File: .github/workflows/deployDev.yml
Comment (alismx) [reply .github/workflows/deployDev.yml:9]: Thanks for calling this out, this was a mistake.

File: frontend/cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json
Comment (alismx) [reply frontend/cypress/stubs/orgSignUp/mappings/api_v1_users-d2c41957-0331-4c20-a597-7002630980ab.json:8]: I am having the darndest time getting regex to work on this Wiremock field. The more I struggle to get it working, the more I think about how Faker is filling in this data and our app is validating the input. I think we'll be ok with the other hard-coded values and accept whatever is here. Thoughts?

File: .github/workflows/deployStg.yml
Comment (nickclyde) [inline .github/workflows/deployStg.yml:112]: Nice addition! Great to have in case we need to push a fix out urgently

File: frontend/cypress.json
Comment (nickclyde) [inline frontend/cypress.json:6]: Nice 👍

File: frontend/cypress/integration/01-organization_sign_up_spec.js
Comment (nickclyde) [inline frontend/cypress/integration/01-organization_sign_up_spec.js:10]: Good stuff 👍 At some point, I'd like to add some mechanism to clean up these randomly created orgs from staging after the tests past, to avoid filling up our Okta instance with a ton of junk groups. But I don't think it should block this PR. Can you create a ticket for that work so we don't lose track?

**Review by nickclyde — APPROVED** [review]

File: .github/workflows/deployStg.yml
Comment (rin-skylight) [reply .github/workflows/deployStg.yml:112]: I'm not so sure I agree with this. If we need to make an urgent repair on something, we can theoretically leverage the Azure deployment slots to do so. Introducing a bypass to a check makes me feel squidgy.

Is there a situation we can think of where a slot-based rollback wouldn't solve an immediate problem, and let us get a fix out with all the proper checks in place?

Comment (alismx) [reply .github/workflows/deployStg.yml:112]: TBH, I don't love bypasses like this either. However, I know we've had issues with fixes taking more time than seems comfortable, and this felt like a generally low-risk job to have this sort of bypass on, which is why I didn't push back.

Overall, I'd love to audit and work to optimize our pipeline instead of adding checks like this.

File: frontend/cypress/integration/01-organization_sign_up_spec.js
Comment (alismx) [reply frontend/cypress/integration/01-organization_sign_up_spec.js:10]: I'll write up those tickets before merging this PR.

File: .github/workflows/deployStg.yml
Comment (alismx) [reply .github/workflows/deployStg.yml:112]: After chatting with Rin, I'm going to remove this skip option. We should strive to keep the integrity of our pipelines intact and not skip verification steps. We also talked about prioritizing work to make our pipelines more efficient to allow for faster deploys, and we have at least one ticket coming up that would do this.

File: frontend/cypress/integration/01-organization_sign_up_spec.js
Comment (alismx) [reply frontend/cypress/integration/01-organization_sign_up_spec.js:10]: https://app.zenhub.com/workspaces/simplereport-eng-health-61127bc8b61cae0013456681/issues/cdcgov/prime-simplereport/3630

**Comment by alismx** [comment]
Will reopen on a new branch that includes the docker changes. Pairing with Nick on an authorization problem on Friday to finish that up.

**Pull request closed without merging**
